src/core/writer.js

```diff
@@ -0,0 +1,164 @@
+/* Copyright 2020 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint no-var: error */
+
+import { bytesToString, escapeString, stringToBytes } from "../shared/util.js";
+import { calculateMD5 } from "./crypto.js";
+import { Dict } from "./primitives.js";
+
+function writeDict(dict, buffer) {
+  buffer += "<<";
+  for (const [key, value] of Object.entries(dict._map)) {
```

Comment:
This may be moot with my other comment addressed (about a correct main/worker-thread split), but please note the `_map` property on a `Dict` instance should be considered *private* and consequently you *really* aren't supposed to be doing this!